### ISSUE TITLE
Unknown import symbol when using __all__

### ISSUE THREAD

**Issue by fabiencelier** [microsoft/pyright#1131]
**Describe the bug**
Pyright reports an "Unknown import symbol" when I think it should not. I am defining a function in a module but re-exporting twice before using it. 

**To Reproduce**

This is a simplified example of what happens in our code. It's still a bit complex but I could not reproduce with less.

Define a function in `a.py`
```python
def example():
    print("test")
```

Import it by name in `b.py` and redefine `__all__`
```
from .a import example as example

__all__ = [example.__name__]
```
In `c.py`, Import it from `b` with `*`:

```python
from .b import *
```
Use it into `d.py`

```python
from .c import example # This is the issue:  "example" is unknown import symbol - Pyright (reportGeneralTypeIssues)
```

This looks like an issue with the re-export because in `c.py` the import works. Also if I don't redefine the `__all__` there is no issue.

**VS Code extension or command-line**
both with version 1.1.82

**Comment by erictraut** [comment]
Pyright does not allow you to use dynamic expressions within an `__all__` expression. It must consist of string literals only. Please refer to [this documentation](https://github.com/microsoft/pyright/blob/master/docs/typed-libraries.md) for details.

**Comment by fabiencelier** [comment]
It can even be replicated with 2 modules:

a.py:
```python
def example():
    print("test")

__all__ = [example.__name__]
```

b.py
```
from .a import *

example() # This is the issue
```

Is this a limitation to simplify Pyright ?

I am not very comfortable with using strings for function names in `__all__`, I always use `function.__name__` because I consider it easier to refactor.

**Comment by erictraut** [comment]
This is a limitation of a static type checker in general. We recently discussed the topic of `__all__` in the typing-sig among all the maintainers of Python static type checkers (pyright, mypy, pyre, pyype, tc.). Our goal is to clarify and standardize which forms of `__all__` are supported by all type checkers. We landed on the proposal that is documented in the link provided above. That's what's currently implemented in pyright. Now that this has become somewhat of an agreed-upon standard among type checkers, it's unlikely to change.

**Comment by fabiencelier** [comment]
Ok, thank you for the clarifications !

Should Pyright issue a warning if a user tries to set `__all__` to something unsupported so the issue is easier to understand ?

**Comment by erictraut** [comment]
That's a reasonable request. I've filed a separate feature request (https://github.com/microsoft/pyright/issues/1133).

**Comment by fabiencelier** [comment]
Would you have a link to this discussion about `__all__` among maintainers of static type checkers or anything "official" following this discussion ?

**Comment by erictraut** [comment]
Here's a link to the thread: https://mail.python.org/archives/list/typing-sig@python.org/thread/YLJPWECBNPD2K4TRIBRIPISNUZJCRREY/
